### ISSUE TITLE
model contributions more in line with tf-slim

### ISSUE THREAD

**Issue by ahundt** [farizrahman4u/keras-contrib#15]
There are a number of good model implementations in third party repositories, and I think keras-contrib may be the most sensible place for them to go. However, each will need an example and a way to train, or a single script that works for all models. For that reason I propose a [tf-slim](https://github.com/tensorflow/models/tree/master/slim) style setup that can eliminate some of that overhead for users and provide more general utility than the [keras/applications](https://github.com/fchollet/keras/blob/master/keras/applications/) folder.

The following may be aspects of a model folder:

 - [ ] have a script equivalent to [train_image_classifier.py](https://github.com/tensorflow/models/blob/master/slim/train_image_classifier.py) explained in the [tf-slim readme](https://github.com/tensorflow/models/tree/master/slim).
 - [ ] train_image_classifier might support both image labeling and semantic segmentation
 - [ ] model test folder, perhaps in a folder `models` placed in https://github.com/farizrahman4u/keras-contrib/tree/master/tests/keras_contrib 

One such example model implementation is [keras-resnet](https://github.com/raghakot/keras-resnet). The author is [interested in submitting a pull request to keras-contrib](https://github.com/raghakot/keras-resnet/issues/32).

I know that [deep-learning-models](https://github.com/fchollet/deep-learning-models) exists but there are benefits keras-contrib could *eventually* provide:

- provide way to train the models from scratch
- community administrators (doesn't rely on availability of fcollet)
- code (or at least no documentation) to convert trained models from other deep learning apis like caffe

Would the ideal place to put such a thing be in `kerascontrib/keras-contrib/models`?

**Comment by the-moliver** [comment]
I have mixed feelings about this. While centralization can be a good thing, I think a model library might go beyond the intended scope of this repository. I had imagined this to be more of a testing ground for cutting edge features from recent literature. Perhaps it would be better to expand on fchollet's deep-learning-models if he's willing to grant permissions? What do others think?

**Comment by farizrahman4u** [comment]
This repo is intended for testing new features (new layers, backend functions, optimizers, etc), as well as examples that use such features. If you have models that use only existing Keras features, it would be better to PR to @fchollet's deep learning models repo.

**Comment by kemaswill** [comment]
We all agree that there should be an official keras repository for the community to contribute new models. Ideally it should be in an separate repo such as @fchollet 's deep learning models repo. However, there will be some models that may use new features which only exist in keras-contrib, so some of the new models MUST be pushed to keras-contri while the others goes to the model repo. Therefore **I think we can just make it simple and use this repo as the only one repo for all models** so that the contributors don't need to wonder which model repo they should pushed their models to.

**Comment by tboquet** [comment]
Would it be possible to use @fchollet 's deep learning models repo and add a folder keras-contrib-models with a readme. It could be explicit enough to advise the user that he should use the keras-contrib repo with these models?
I could also be possible to explain how to set up a virtualenv and provide a Dockerfile so it will be easy to work in a clean environment?
I guess the reviewers from the keras-contrib repo should review the models associated with the features merged here.
What do you think?

**Comment by ahundt** [comment]
Allow me to tweak my suggestion so that as you specified, I'll be proposing a specific new feature as well as examples of using that feature:

1) Define a standard class to create the most prominent published models of various types that can generate each model and configure the model including:
    - single class, multi class, segmentation, and instance segmentation, where applicable
    - network size for various hardware and applications (i.e. the many different scales of resnet/densenet that are available, for example)
    - provide callbacks or attachment points in this API for appropriate attachment points to add/train additional data
2) Create a new training class (or extend the existing one) to meet needs not met by the current one and automatically handle the standard class in (1).
    - Provide a built in multi-gpu training queue
    - Support deeply supervised learning
    - Provide hyperparameter search on these models (may be part of model API).
    - Automate training of a set of models/configurations for evaluation
3) Provide a simple applications folder app that initializes these utilities and demonstrates their use:
    - trains, saves, and evaluates models from the command line
    - loads and run a selection of datasets
    - provide a simple folder layout you can use to supply your own dataset

As requested, I believe this meets the standard @farizrahman4u is looking for of testing new features, as well as creating examples of using such features. However, this isn't a complete design yet, just a request for comment.

**Comment by ahundt** [comment]
In other words, I'm trying to bootstrap some of the tools needed for applied R&D in a way that is much more disciplined than the quirky model/weight libraries in other deep learning projects.

**Comment by tboquet** [comment]
I see, in my opinion (that could be wrong) it would definitely be possible to impose a simple style guide for the examples and maybe normalize the existing ones. The whole applied research pipeline is really interesting and important but it seems hard to integrate in this repo (API, weights management, experiment planning, etc.). I imagine it could be done in a separate project.
